pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go

```diff
@@ -0,0 +1,150 @@
+)
+
+var _ OperationGenerator = &operationGenerator{}
+
+type operationGenerator struct {
+
+	// recorder is used to record events in the API server
+	recorder record.EventRecorder
+}
+
+// NewOperationGenerator is returns instance of operationGenerator
+func NewOperationGenerator(recorder record.EventRecorder) OperationGenerator {
+
+	return &operationGenerator{
+		recorder: recorder,
+	}
+}
+
+// OperationGenerator interface that extracts out the functions from operation_executor to make it dependency injectable
+type OperationGenerator interface {
+	// Generates the RegisterPlugin function needed to perform the registration of a plugin
+	GenerateRegisterPluginFunc(
+		pluginToRegister cache.PluginInfo,
+		pluginHandler cache.PluginHandler,
+		actualStateOfWorldUpdater ActualStateOfWorldUpdater) (func() error, error)
+
+	// Generates the UnregisterPlugin function needed to perform the unregistration of a plugin
+	GenerateUnregisterPluginFunc(
+		pluginToDeRegister cache.PluginInfo,
+		pluginHandler cache.PluginHandler,
+		actualStateOfWorldUpdater ActualStateOfWorldUpdater) (func() error, error)
+}
+
+func (og *operationGenerator) GenerateRegisterPluginFunc(
+	pluginToRegister cache.PluginInfo,
+	pluginHandler cache.PluginHandler,
+	actualStateOfWorldUpdater ActualStateOfWorldUpdater) (func() error, error) {
+
+	registerPluginFunc := func() error {
+		// client, conn, err := dial(pluginToRegister.SocketPath, 10*time.Second)
+		// if err != nil {
+		// 	return nil, fmt.Errorf("dial failed at socket %s, err: %v", pluginToRegister.SocketPath, err)
+		// }
+		// defer conn.Close()
+		// calls handler callback to verify registration request
+		if err := pluginHandler.ValidatePlugin(pluginToRegister.Name, pluginToRegister.Endpoint, pluginToRegister.SupportedVersions, pluginToRegister.FoundInDeprecatedDir); err != nil {
+			return fmt.Errorf("pluginHandler.ValidatePluginFunc failed")
+		}
+		// We add the plugin to the actual state of world cache before calling a plugin consumer's Register handle
+		// so that if we receive a delete event during Register Plugin, we can process it as a DeRegister call.
+		actualStateOfWorldUpdater.AddOrUpdatePlugin(cache.PluginInfo{
+			Name:                 pluginToRegister.Name,
+			Type:                 pluginToRegister.Type,
+			SocketPath:           pluginToRegister.SocketPath,
+			FoundInDeprecatedDir: pluginToRegister.FoundInDeprecatedDir,
+			Endpoint:             pluginToRegister.Endpoint,
+			SupportedVersions:    pluginToRegister.SupportedVersions,
+		})
+		if err := pluginHandler.RegisterPlugin(pluginToRegister.Name, pluginToRegister.Endpoint, pluginToRegister.SupportedVersions); err != nil {
+			return fmt.Errorf("pluginHandler.RegisterPluginFunc failed")
+		}
+		// // Notify is called after register to guarantee that even if notify throws an error Register will always be called after validate
+		// if err := og.notifyPlugin(client, true, ""); err != nil {
```

Comment:
We should still do this. Essentially this is a call back to the plugin to say, hey the handler for your type says you are not a valid plugin. So heads up, registration failed.